File: test/ctest/case/query_api.js
'use strict';

function init(input_str) {
    Chain.store('init', input_str);
    return '';
}

function main(input_str) {
    if (input_str === 'load') {
        Chain.store('load', JSON.stringify(Chain.delegateQuery(Chain.thisAddress, input_str)));
    }

    if (input_str === 'store') {
        Chain.store('store', JSON.stringify(Chain.delegateQuery(Chain.thisAddress, input_str)));
    }

    if (input_str === 'del') {
        Chain.store('del', JSON.stringify(Chain.delegateQuery(Chain.thisAddress, input_str)));
    }

    if (input_str === 'getBlockHash') {
        Chain.store('getBlockHash', JSON.stringify(Chain.delegateQuery(Chain.thisAddress, input_str)));
    }

    if (input_str === 'tlog') {
        Chain.store('tlog', JSON.stringify(Chain.delegateQuery(Chain.thisAddress, input_str)));
    }

    if (input_str === 'getValidators') {
        Chain.store('getValidators', JSON.stringify(Chain.delegateQuery(Chain.thisAddress, input_str)));
    }

    if (input_str === 'getAccountMetadata') {
        Chain.store('getAccountMetadata', JSON.stringify(Chain.delegateQuery(Chain.thisAddress, input_str)));
    }

    return '';
}

function query(input_str) {
    if (input_str === 'load') {
        return Chain.load('init');
    }

    if (input_str === 'store') {
        return Chain.store('init', 'hello');
    }

    if (input_str === 'del') {
        return Chain.del('init');
    }

    if (input_str === 'getBlockHash') {
        return Chain.getBlockHash(0);
    }

    if (input_str === 'tlog') {
        return Chain.tlog(0);
    }

    if (input_str === 'getValidators') {
        return JSON.stringify(Chain.getValidators());
    }

    if (input_str === 'getAccountMetadata') {
        return Chain.getAccountMetadata(Chain.thisAddress, 'init');
    }

    return '';
}

File: test/ctest/case/query_api.cpp
#include "ctest.h"
#include "gtest/gtest.h"

class QueryAPI : public testing::Test{
protected:
	void test();
	bumo::CTest ctest_;
};

TEST_F(QueryAPI, test){ test(); }

void QueryAPI::test(){
	//create contract
	bumo::CTest::Param param;
	param.balance = "10000000000";
	param.fee_limit = "20000000000";
	param.gas_price = "1000";
	param.input = "init";
	EXPECT_EQ(ctest_.Create(__FILE__, param), 0);

	param.input = "load";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	EXPECT_EQ(ctest_.Load("load"), "{\"result\":\"init\"}");

	param.input = "store";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	EXPECT_EQ(ctest_.Load("store"), "{\"error\":true}");

	param.input = "del";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	EXPECT_EQ(ctest_.Load("del"), "{\"error\":true}");

	param.input = "getBlockHash";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	EXPECT_NE(ctest_.Load("getBlockHash"), "");
	EXPECT_NE(ctest_.Load("getBlockHash"), "{\"error\":true}");

	param.input = "tlog";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	EXPECT_EQ(ctest_.Load("tlog"), "{\"error\":true}");

	param.input = "getValidators";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	EXPECT_NE(ctest_.Load("getValidators"), "");
	EXPECT_NE(ctest_.Load("getValidators"), "{\"error\":true}");

	param.input = "getAccountMetadata";
	EXPECT_EQ(ctest_.Paycoin(param), 0);
	EXPECT_EQ(ctest_.Load("getAccountMetadata"), "{\"result\":\"init\"}");
}
